mimxrt: Initial documentation for the mimxrt port.

robert-hh opened the pull request:
While the mimxrt port is made as compatible as possible to the mainline of ports, some aspects are slightly different. These documentation pages address some of them. If consists mainly of a quickref, the introduction and the 'general' pages. More to follow.
Note: I2C is already mentioned, even if it not yet merged. And SDCard as well, which is in good shape and under test and refinement now.

robert-hh committed:
mimxrt: Doc pages for the mimxrt port.

dpgeorge commented:
Is this ready for final review, or do you want to wait until other features are implemented?

dpgeorge commented on docs/mimxrt/general.rst:
```diff
@@ -0,0 +1,94 @@
+
+Note
+    Most of the controllers do not have internal flash memory. Therefore
+    their flash capacity is dependent on an external flash chip!
```
In general I suggest that exclamation marks are not used in documentation unless really warranted.

robert-hh replied:
OK. Will be changed.

robert-hh commented:
> Is this ready for final review, or do you want to wait until other features are implemented?

That is definitely an interim state, but I made it to show the Pin assignments of Devices like UART, SPI, I2C and PWM to pins. These are documented for Teensy, but not for the MIMXRT EVK boards. It contains some documentation about features in progress to get publish, like SDCard and PWM, and also the tiny DHT support addition which I have ready but did not send a PR yet. I'm preparing another "various" PR for collected small additions.
So yes, publishing it is helpful, but there will be updates and extensions, if new features are added, like CAN, I2S, Watchdog, extended Flash, ... The devices has still quite a  few options to offer.

IhorNehrutsa commented on docs/mimxrt/machine.QECNT.rst:
```diff
@@ -0,0 +1,377 @@
+      - *filter_ns*\=value. Specifies a ns-value for the minimal time a signal has to be stable
+        at the input to be recognized. The code does the best effort to configure the filter.
+        The largest value is 20400 for the i.MXRT102x and 17000 for the i.MXRT105x/i.MXRT106x
+        (1000000000 * 2550 * 4 / CPU_CLK). A value of 0 sets the filter sets the filter off.
```
sets the filter sets the filter - duplication

IhorNehrutsa commented on docs/mimxrt/machine.QECNT.rst:
```diff
@@ -0,0 +1,377 @@
+   Specifies, that the *handler* is called when the respective *event* happens.
+
+   *event* may be:
+    - Encoder.COMPARE Triggered when the position counter matches the compare value.
+    - Encoder.ROLL_OVER Triggered when the position counter rolls over from the highest
+      to the lowest value.
+    - Encoder.ROLL_UNDER Triggered when the position counter rolls under from the lowest
+      to the highest value.
```
It seems to me that it would be more correct to call
Encoder.IRQ_COMPARE
Encoder.IRQ_ROLL_OVER
Encoder.IRQ_ROLL_UNDER
like the
Pin.IRQ_FALLING
Pin.IRQ_RISING

robert-hh replied:
Being a lazy typer, I prefer shorter names.

IhorNehrutsa replied:
In the vast majority of cases, MicroPython uses IRQ_XXX.
https://docs.micropython.org/en/latest/search.html?q=irq&check_keywords=yes&area=default#

robert-hh commented:
@dpgeorge Does this Doc part have to be re-structured. I'm not sure about the extended PWM and Encoder section. They may have to be linked from another place (beside the Encoder PR not being merged yet).

robert-hh committed:
mimxrt: Add a chapter for network.LAN.

And rearrange the files and links.
Remove for now the file for the Encoder, such that the documentation
matches the actual port.

robert-hh committed:
mimxrt: Add the Pin settings for the MIMXRT1015_EVK board.

And fix a formatting error in machine.PWM.rst.

robert-hh commented:
@dpgeorge Good day Damien. Meanwhile, it would be good to have this documentation PR merged, creating the baseline for the MIMXRT port's documentation. That would allow other contributors like Mike Teachman with his I2S class or alphaFred with the generic bootloader to extend it.

robert-hh committed:
mimxrt: Add the second CS pin info.

Only few boards support it.

robert-hh committed:
mimxrt: Add a I2S section to the Quickref.

robert-hh committed:
mimxrt: Add notes for the Olimex RT1010Py module.

robert-hh committed:
mimxrt: Edits to the PWM doc.

robert-hh committed:
mimxrt: Change SPI1 pins of Teensy 4.1.

After it was changed in the code.

robert-hh committed:
mimxrt/qickref: Fix two typos and one value.

dpgeorge commented:
In this PR there are `machine.PWMmimxrt.rst` and `network.LANmimxrt.rst` files.  This does not look right, the API for PWM and LAN should match other ports and so the mimxrt should just use `machine.PWM.rst` and `network.LAN.rst`.  Additional information about PWM/LAN specific to the mimxrt port should go in the following locations (depending on which makes more sense):
1. existing `machine.PWM.rst`/`network.LAN.rst` docs in mimxrt-specific sections
2. mimxrt quickref
3. mimxrt tutorial

robert-hh commented:
Thank you for all the review work. You cleaned up many PRs piling up since long. I recall some like the one about the "low hanging fruits", which seemed out of reach for me when reading it the first time.
I'll work on the suggested changes.

robert-hh commented:
@dpgeorge About moving network.LANmimxrt.rst file: There is no network.LAN,rst file. There is a network.rst file, describing an abstract class, and a set of specific files for CC3, WIZNET5K, WLAN, WLANWiPy.  The LANmimxrt.rst file follows that structure.

robert-hh committed:
mimxrt: Move the PWM documentation back to quickref.rst.

Since it contains all the pin assignment tables, it fits better there.

dpgeorge commented:
> About moving network.LANmimxrt.rst file: There is no network.LAN,rst file

OK, then this network.LANmimxrt.rst file can be renamed to network.LAN.rst and form the basis of the docs for that class.

robert-hh committed:
mimxrt: Change network.LANmimxrt.rst to network.LAN.rst.

Editorial change: Move the MIMXRT specific sentence to the end of
the section.

robert-hh commented:
Done & updated in the repository.

dpgeorge commented:
Squashed and merged in 7a447e08b23f47fac6ebcbe5d60c4e1d0fe0d157, with minor changes:
- removed unused block diagram image
- reduced size of teensy 4.1 image
- fixed download link
- fixed typos
- fixed and adjusted formatting

One thing I would say is that the quickref is too long.  I think the tables of pins (eg for PWM) could go elsewhere, eg in a specific file with pinouts for each board.  But that's not a critical thing.

Anyway, this is a good addition, thank you!

robert-hh commented:
Thank you very much for merging. I'll split the quickref up with the next change.
